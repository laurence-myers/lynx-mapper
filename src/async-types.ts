/**
 * These types are used by the AsyncObjectMapper. Here's the magic which makes
 *  object mapper schemas type safe and complete. You generally won't need to
 *  make use of these types within your own code; just `ObjectMapper` should be
 *  enough.
 *
 * @module
 */

import {
  AllowInputKeyIfInputCanExtendOutput,
  AllowOmitIfOptional,
  OptionalArgIfUndefined,
} from "./types.ts";

/**
 * A function that takes some input object, and an optional context object, and returns a
 *  promise of an output. This is used as part of an {@linkcode AsyncObjectMapperSchema}.
 * @param input The input object.
 * @param context The context object. If the `TContext` type parameter is `undefined`, you can omit this argument.
 * @returns The output value, OR the special symbol {@linkcode OmitProperty}`, indicating that the property should be omitted
 *   completely.
 */
export interface AsyncMapperFunction<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined,
  TOutputKey extends keyof TOutput = keyof TOutput,
> {
  <TInputSubset extends TInput, TContextSubset extends TContext>(
    input: TInputSubset,
    context: OptionalArgIfUndefined<TContextSubset>,
  ): Promise<TOutput[TOutputKey] | AllowOmitIfOptional<TOutput, TOutputKey>>;
}

/**
 * A mapper function, or input property name, used in an {@linkcode AsyncObjectMapperSchema}.
 *
 * In an {@linkcode AsyncObjectMapperSchema}, each property value can be either:
 * - An {@linkcode AsyncMapperFunction}
 * - A property name from the input object. The property value must be compatible
 *   with the output property.
 */
export type AsyncMapperSchemaValue<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined,
  TOutputKey extends keyof TOutput = keyof TOutput,
> =
  | AsyncMapperFunction<TInput, TOutput, TContext, TOutputKey>
  | AllowInputKeyIfInputCanExtendOutput<TInput, TOutput[TOutputKey]>;

/**
 * An object, where every property name must match a property name in the desired output type.
 *  Every property value must be a {@linkcode AsyncMapperSchemaValue}.
 * ```
 */
export type AsyncObjectMapperSchema<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined = undefined,
> = {
  [TOutputKey in keyof TOutput]-?: AsyncMapperSchemaValue<
    TInput,
    TOutput,
    TContext,
    TOutputKey
  >;
};

/**
 * A callable function, equivalent to calling {@linkcode AsyncObjectMapper#map}.
 *  It also exposes {@linkcode AsyncObjectMapperFunction#schema} as a readonly property.
 */
export interface AsyncObjectMapperFunction<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined = undefined,
> {
  (value: TInput, context: OptionalArgIfUndefined<TContext>): Promise<TOutput>;

  readonly schema: AsyncObjectMapperSchema<TInput, TOutput, TContext>;
}
