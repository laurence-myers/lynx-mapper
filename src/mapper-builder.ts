type OptionalArgIfUndefined<T> = T extends undefined ? T | void : T;

export interface MapperFunction<TInput, TOutput, TContext> {
  (input: TInput, context: OptionalArgIfUndefined<TContext>): TOutput;
}

/**
 * Needs to know the initial input and context from the outer mapper. Can be done by changing ObjectMapper to accept
 * a callback function, whose first argument is a MapperBuilder.
 *
 * MapperBuilder could store an array of transformation functions. Building could pipe the inputs through each
 * function. (Needs at least one function; could default to "identity", i.e. return the input.)
 *
 * Should be immutable: each build returns a clone. (For reusing schemas)
 */

export class MapperBuilder<TInput, TOutput, TContext> {
  constructor(protected readonly transforms: Function[] = []) {}

  context(getter: <TContext>(input: TInput, context: TContext) => TOutput) {
    return new MapperBuilder<TInput, TOutput, TContext>(
      this.transforms.concat(getter)
    );
  }

  get(getter: <TOutput>(input: TInput, context: TContext) => TOutput) {
    return new MapperBuilder<TInput, TOutput, TContext>(
      this.transforms.concat(getter)
    );
  }

  build(): MapperFunction<TInput, TOutput, TContext> {
    return (input: TInput, context: OptionalArgIfUndefined<TContext>) => {
      for (const transform of this.transforms) {
        input = transform(input, context);
        // How to support transforming "context"?
      }
      return input as unknown as TOutput;
    };
  }
}
