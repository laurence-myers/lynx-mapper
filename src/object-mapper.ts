export interface MapperFunction<
  TInput extends object,
  TOutput,
  TContext extends object | undefined
> {
  (input: TInput, context: OptionalArgIfUndefined<TContext>): TOutput;
}

type AllowInputKeyIfInputCanExtendOutput<TInput, TOutputValue> = {
  [TInputKey in keyof TInput]: TInput[TInputKey] extends TOutputValue
    ? TInputKey
    : never;
}[keyof TInput];

type OptionalArgIfUndefined<T> = T extends undefined ? T | void : T;

/**
 * Every property name must match a property name in the desired output type.
 * Every property value must be a function, that accepts the whole input value,
 * and returns a single value conforming to the output type's property type.
 *
 * A property value can _also_ be a string, matching a property name on the
 * input type, _if_ the input property value has a type acceptable to the
 * output property.
 * For example:
 *
 * ```
 * interface Input {
 *   inString: string;
 *   inNumber: number;
 *   inOptionalString: string | undefined;
 * }
 *
 * interface Output {
 *   outString: string;
 * }
 *
 * new ObjectMapper<Input, Output>({
 *   outString: "inString" // This is okay, since `Input["inString"]` has the same type as `Output["outString"]`
 * });
 *
 * new ObjectMapper<Input, Output>({
 *   outString: "inNumber" // This will error. `inNumber` has type `number`, so can't be used to populate `outString` which requires a `string`.
 * });
 *
 * new ObjectMapper<Input, Output>({
 *   outString: "inOptionalString" // This will error. `inOptionalString` can be `undefined`, which isn't compatible with `string` (when strict null checks are on).
 * }, undefined);
 * ```
 *
 * The mapper function accepts a "context", which can store extra data, alongside the specific object you're mapping.
 * Pass `undefined` if you don't need any extra data. You can also omit the argument.
 */
export type ObjectMapperSchema<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined
> = {
  [K in keyof TOutput]-?:
    | MapperFunction<TInput, TOutput[K], TContext>
    | AllowInputKeyIfInputCanExtendOutput<TInput, TOutput[K]>;
};

interface ObjectMapperFunctionBeingBuilt<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined = undefined
> {
  (value: TInput, context: OptionalArgIfUndefined<TContext>): TOutput;

  schema: ObjectMapperSchema<TInput, TOutput, TContext>;
}

export interface ObjectMapperFunction<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined = undefined
> {
  (value: TInput, context: OptionalArgIfUndefined<TContext>): TOutput;

  readonly schema: ObjectMapperSchema<TInput, TOutput, TContext>;
}

/**
 * Convert from one type of object to another.
 */
export class ObjectMapper<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined = undefined
> {
  /**
   * Used for mapping one array to another array.
   */
  public static array<
    TInput extends object,
    TContext extends object | undefined,
    TNestedInput extends object,
    TNestedOutput extends object,
    TNestedContext extends object | undefined
  >(
    mapper: MapperFunction<TNestedInput, TNestedOutput, TNestedContext>,
    inputGetter: (
      input: TInput,
      context: OptionalArgIfUndefined<TContext>
    ) => readonly TNestedInput[],
    contextFactory: (
      input: TInput,
      context: OptionalArgIfUndefined<TContext>
    ) => TNestedContext
  ): MapperFunction<TInput, TNestedOutput[], TContext> {
    return function arrayMapper(input, context) {
      const nestedInput = inputGetter(input, context);
      const nestedContext = contextFactory(input, context);
      return nestedInput.map((value) =>
        mapper(value, nestedContext as OptionalArgIfUndefined<TNestedContext>)
      );
    };
  }

  /**
   * Returns the input value.
   */
  public static identity<T>(value: T): T {
    return value;
  }

  public static nested<
    TInput extends object,
    TContext extends object | undefined,
    TNestedInput extends object,
    TNestedOutput extends object,
    TNestedContext extends object | undefined
  >(
    objectMapper: ObjectMapper<TNestedInput, TNestedOutput, TNestedContext>,
    inputGetter: (
      input: TInput,
      context: OptionalArgIfUndefined<TContext>
    ) => TNestedInput,
    contextFactory: (
      input: TInput,
      context: OptionalArgIfUndefined<TContext>
    ) => TNestedContext
  ): MapperFunction<TInput, TNestedOutput, TContext> {
    return function nestedObjectMapper(input, context) {
      const nestedInput = inputGetter(input, context);
      const nestedContext = contextFactory(input, context);
      return objectMapper.map(
        nestedInput,
        nestedContext as OptionalArgIfUndefined<TNestedContext>
      );
    };
  }

  public static null(): null {
    return null;
  }

  public static undefined(): undefined {
    return undefined;
  }

  constructor(
    public readonly schema: ObjectMapperSchema<TInput, TOutput, TContext>
  ) {}

  // Unsafe stuff happens here
  map(value: TInput, context: OptionalArgIfUndefined<TContext>): TOutput {
    const output: Record<string, any> = {};
    for (const [key, getterOrString] of Object.entries(this.schema)) {
      if (typeof getterOrString === "string") {
        output[key] = (value as any)[getterOrString];
      } else {
        output[key] = (
          getterOrString as MapperFunction<TInput, TOutput, TContext>
        )(value, context);
      }
    }
    return output as TOutput;
  }

  toFunction(): ObjectMapperFunction<TInput, TOutput, TContext> {
    const func: ObjectMapperFunctionBeingBuilt<TInput, TOutput, TContext> = (
      value,
      context
    ) => this.map(value, context as OptionalArgIfUndefined<TContext>);
    func.schema = this.schema;
    return func;
  }
}
