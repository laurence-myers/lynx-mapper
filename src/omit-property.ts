export const OmitProperty: unique symbol = Symbol("OmitProperty");
/**
 * Use this symbol to tell an ObjectMapper to exclude the property completely from the
 *  output object. This is useful when you must distinguish between `"property" in output`
 *  vs `output.property === undefined`.
 */
export type OmitProperty = typeof OmitProperty;
