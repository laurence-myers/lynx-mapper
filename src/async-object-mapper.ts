import { OmitProperty } from "./omit-property.ts";
import type { OptionalArgIfUndefined } from "./types.ts";
import {
  AsyncMapperFunction,
  AsyncMapperSchemaValue,
  AsyncObjectMapperFunction,
  AsyncObjectMapperSchema,
} from "./async-types.ts";

/**
 * An internal type used by {@linkcode AsyncObjectMapper.toFunction()}
 *
 * @private
 */
interface AsyncObjectMapperFunctionBeingBuilt<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined = undefined,
> {
  (value: TInput, context: OptionalArgIfUndefined<TContext>): Promise<TOutput>;

  schema: AsyncObjectMapperSchema<TInput, TOutput, TContext>;
}

/**
 * Used to determine if a {@linkcode MapperSchemaValue} is a mapper function,
 *  or a string (representing a property name on some output type).
 *
 * @private
 */
function isAsyncMapperFunction<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined = undefined,
>(
  value: AsyncMapperSchemaValue<TInput, TOutput, TContext>,
): value is AsyncMapperFunction<TInput, TOutput, TContext> {
  return typeof value === "function";
}

/**
 * Convert from one type of object to another.
 *
 * Instantiate an instance with a {@linkcode AsyncObjectMapperSchema}. You must pass two type parameters:
 *
 * - `TInput`: the type of the input object
 * - `TOutput`: the type of the desired output object
 *
 * You can pass an optional third type parameter:
 *
 * - `TContext`: an object with any additional data or functions. Useful when mapping multiple
 *   objects with some shared state, like a "now" timestamp.
 *
 * Invoke the mapper with {@linkcode AsyncObjectMapper#map}.
 *
 * If you want a plain (unbound) function, you can call {@linkcode AsyncObjectMapper#toFunction}.
 *
 * There is a convenience method {@linkcode AsyncObjectMapper#array}, for mapping some iterable to an array.
 *
 * @group async
 */
export class AsyncObjectMapper<
  TInput extends object,
  TOutput extends object,
  TContext extends object | undefined = undefined,
> {
  /**
   * Create an AsyncObjectMapper factory function. Invoke it immediately,
   *  with an object mapper schema, to create an AsyncObjectMapper instance.
   *
   * We use this approach to trick TypeScript into requiring an exact type
   *  to be passed in. That is, the object mapper schema must only have
   *  properties that exist in the output type, and no additional properties.
   */
  public static create<
    TInput extends object,
    TOutput extends object,
    TContext extends object | undefined = undefined,
  >() {
    type TDesiredSchema = AsyncObjectMapperSchema<TInput, TOutput, TContext>;
    return function <
      TActualSchema extends
        & TDesiredSchema
        & {
          [
            P in Exclude<
              keyof TActualSchema,
              keyof TDesiredSchema
            >
          ]: never;
        },
    >(
      schema: TActualSchema,
    ): AsyncObjectMapper<TInput, TOutput, TContext> {
      return new AsyncObjectMapper<TInput, TOutput, TContext>(schema);
    };
  }

  /**
   * For faster runtime performance, the object mapper schema is converted to
   *  a Map instance.
   * @private
   */
  protected readonly schemaMap: Map<
    keyof TOutput,
    AsyncMapperSchemaValue<TInput, TOutput, TContext>
  >;

  protected constructor(
    /**
     * Defines how to populate property on the output type.
     */
    public readonly schema: AsyncObjectMapperSchema<TInput, TOutput, TContext>,
  ) {
    this.schemaMap = new Map<
      keyof TOutput,
      AsyncMapperSchemaValue<TInput, TOutput, TContext>
    >(Object.entries(schema) as [
      keyof TOutput,
      AsyncMapperSchemaValue<TInput, TOutput, TContext>,
    ][]);
  }

  /**
   * Map multiple input objects from some iterable, and return an
   *  array of output objects.
   */
  array(
    input: Iterable<TInput>,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput[]>;
  /**
   * Map multiple input objects from some iterable, and return an
   *  array of output objects.
   *
   * If the input is `null`, it will be returned as-is.
   */
  array(
    input: Iterable<TInput> | null,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput[] | null>;
  /**
   * Map multiple input objects from some iterable, and return an
   *  array of output objects.
   *
   * If the input is `undefined`, it will be returned as-is.
   */
  array(
    input: Iterable<TInput> | undefined,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput[] | undefined>;
  /**
   * Map multiple input objects from some iterable, and return an
   *  array of output objects.
   *
   * If the input is `null` or `undefined`, it will be returned as-is.
   */
  array(
    input: Iterable<TInput> | null | undefined,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput[] | null | undefined>;
  /**
   * Map multiple input objects from some iterable, and return an
   *  array of output objects.
   *
   * If the input is `null` or `undefined`, it will be returned as-is.
   */
  async array(
    input: Iterable<TInput> | null | undefined,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput[] | null | undefined> {
    if (input === undefined || input === null) {
      return input;
    } else if (Array.isArray(input)) {
      // This approach might be faster than using the iterator protocol ("for of" loop)
      return Promise.all(input.map((item) => this.map(item, context)));
    } else {
      const output = [];
      for (const item of input) {
        output.push(await this.map(item, context));
      }
      return output;
    }
  }

  /**
   * Maps an input object to an output object.
   *
   * It does so by iterating each property in the object schema,
   *  and invoking the property's mapping function, passing the input and context.
   */
  map(
    input: TInput,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput>;
  /**
   * Maps an input object to an output object.
   *
   * It does so by iterating each property in the object schema,
   *  and invoking the property's mapping function, passing the input and context.
   *
   * If {@linkcode input} is `null`, it will be returned as-is.
   */
  map(
    input: TInput | null,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput | null>;
  /**
   * Maps an input object to an output object.
   *
   * It does so by iterating each property in the object schema,
   *  and invoking the property's mapping function, passing the input and context.
   *
   * If {@linkcode input} is `undefined`, it will be returned as-is.
   */
  map(
    input: TInput | undefined,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput | undefined>;
  /**
   * Maps an input object to an output object.
   *
   * It does so by iterating each property in the object schema,
   *  and invoking the property's mapping function, passing the input and context.
   *
   * If {@linkcode input} is `null` or `undefined`, it will be returned as-is.
   */
  map(
    input: TInput | null | undefined,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput | null | undefined>;
  /**
   * Maps an input object to an output object.
   *
   * It does so by iterating each property in the object schema,
   *  and invoking the property's mapping function, passing the input and context.
   *
   * If {@linkcode input} is `null` or `undefined`, it will be returned as-is.
   */
  async map(
    input: TInput,
    context: OptionalArgIfUndefined<TContext>,
  ): Promise<TOutput> {
    if (input === null || input === undefined) {
      return input;
    }
    // Unsafe stuff happens here
    const output: Record<string, unknown> = {};
    for (const [key, getterOrString] of this.schemaMap) {
      if (!isAsyncMapperFunction(getterOrString)) {
        output[key as string] = input[getterOrString];
      } else {
        const mappedValue = await getterOrString(input, context);
        if (mappedValue !== OmitProperty) {
          output[key as string] = mappedValue;
        }
      }
    }
    return output as TOutput;
  }

  /**
   * Wrap this instance in a function, with a `schema` property.
   */
  toFunction(): AsyncObjectMapperFunction<TInput, TOutput, TContext> {
    const func: AsyncObjectMapperFunctionBeingBuilt<TInput, TOutput, TContext> =
      (
        value,
        context,
      ) => this.map(value, context);
    func.schema = this.schema;
    return func;
  }
}
