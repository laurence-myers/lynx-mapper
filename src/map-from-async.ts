import { OmitProperty } from "./omit-property.ts";

/**
 * Provides convenience functions for object mappers.
 *
 * Use these to map to constant values, or to omit the property.
 *
 * @category runtime
 */
export const mapFromAsync = {
  /**
   * Returns a function that always returns a constant value.
   *
   * ```ts
   * const objectMapperSchema = {
   *   out: mapFrom.constant('Hello'),
   * }
   * ```
   */
  constant<TValue>(value: TValue): () => Promise<TValue> {
    return function constant(): Promise<TValue> {
      return Promise.resolve(value);
    };
  },

  /**
   * Returns the symbol {@linkcode OmitProperty}, which will tell the
   * mapper to omit the property from the output object.
   *
   * ```ts
   * const objectMapperSchema = {
   *   out: mapFrom.omit,
   * }
   * ```
   */
  omit(): Promise<OmitProperty> {
    return Promise.resolve(OmitProperty);
  },

  /**
   * Always map the output value to `null`
   *
   * ```ts
   * const objectMapperSchema = {
   *   out: mapFrom.null,
   * }
   * ```
   */
  null(): Promise<null> {
    return Promise.resolve(null);
  },

  /**
   * Always map the output value to `undefined`
   *
   * ```ts
   * const objectMapperSchema = {
   *   out: mapFrom.undefined,
   * }
   * ```
   */
  undefined(): Promise<undefined> {
    return Promise.resolve(undefined);
  },
};
