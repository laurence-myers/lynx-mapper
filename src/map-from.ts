import { OmitProperty } from "./omit-property.ts";

export const mapFrom = {
  constant<TValue>(value: TValue): () => TValue {
    return function constant(): TValue {
      return value;
    };
  },
  omit(): OmitProperty {
    return OmitProperty;
  },
  null(): null {
    return null;
  },
  undefined(): undefined {
    return undefined;
  },
};
