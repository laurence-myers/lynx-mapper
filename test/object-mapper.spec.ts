import { ObjectMapper } from "../src/object-mapper";

function stringCounter(start = 0): () => string {
  let i = start;
  return () => {
    return `dummy-string-${++i}`;
  };
}

describe(`ObjectMapper`, () => {
  interface Input {
    inString: string;
    inStringNullable: string | null;
    inStringNullableUndefined: string | null | undefined;
    inStringOptional?: string;
    inStringOptionalNullable?: string | null;
    inStringOptionalNullableUndefined?: string | null | undefined;
    inStringUndefined: string | undefined;
  }

  interface Output {
    outString: string;
    outStringOptional?: string;
    outStringUndefined: string | undefined;
    outStringNullable: string | null;
    outStringOptionalNullable?: string | null;
    outStringNullableUndefined: string | null | undefined;
    outStringOptionalNullableUndefined?: string | null | undefined;
  }

  it(`can map values using strings for property names on the input object`, () => {
    // Setup
    const objectMapper = new ObjectMapper<Input, Output>({
      outString: "inString",
      outStringNullable: "inStringNullable",
      outStringNullableUndefined: "inStringNullableUndefined",
      outStringOptional: "inStringOptional",
      outStringOptionalNullable: "inStringOptionalNullable",
      outStringOptionalNullableUndefined: "inStringOptionalNullableUndefined",
      outStringUndefined: "inStringUndefined",
    });

    const nextString = stringCounter();
    const input: Input = {
      inString: nextString(),
      inStringNullable: nextString(),
      inStringNullableUndefined: nextString(),
      inStringOptional: nextString(),
      inStringOptionalNullable: nextString(),
      inStringOptionalNullableUndefined: nextString(),
      inStringUndefined: nextString(),
    };
    const context = undefined;

    // Execute
    const output = objectMapper.map(input, context);

    // Verify
    expect(output).toMatchSnapshot();
  });

  it(`can map values using functions`, () => {
    // Setup
    const objectMapper = new ObjectMapper<Input, Output>({
      outString: (input) => input.inString,
      outStringNullable: (input) => input.inStringNullable,
      outStringNullableUndefined: (input) => input.inStringNullableUndefined,
      outStringOptional: (input) => input.inStringOptional,
      outStringOptionalNullable: (input) => input.inStringOptionalNullable,
      outStringOptionalNullableUndefined: (input) =>
        input.inStringOptionalNullableUndefined,
      outStringUndefined: (input) => input.inStringUndefined,
    });

    const nextString = stringCounter();
    const input: Input = {
      inString: nextString(),
      inStringNullable: nextString(),
      inStringNullableUndefined: nextString(),
      inStringOptional: nextString(),
      inStringOptionalNullable: nextString(),
      inStringOptionalNullableUndefined: nextString(),
      inStringUndefined: nextString(),
    };
    const context = undefined;

    // Execute
    const output = objectMapper.map(input, context);

    // Verify
    expect(output).toMatchSnapshot();
  });

  it(`can map values using functions, with an additional context object`, () => {
    // Setup
    const objectMapper = new ObjectMapper<
      Pick<Input, "inString">,
      Pick<Output, "outString">,
      { inStringOverride: string }
    >({
      outString: (_, context) => context.inStringOverride,
    });

    const nextString = stringCounter();
    const input: Pick<Input, "inString"> = {
      inString: nextString(),
    };
    const context = {
      inStringOverride: nextString(),
    };

    // Execute
    const output = objectMapper.map(input, context);

    // Verify
    expect(output.outString).not.toEqual(input.inString);
    expect(output.outString).toEqual(context.inStringOverride);
  });

  it(`cannot use the name of a property where the input type is a superset of the output type`, () => {
    // Setup
    new ObjectMapper<Input, Output>({
      // @ts-expect-error TS2322 The input value can't be `undefined` if the output value is `string`
      outString: "inStringUndefined",
      // @ts-expect-error TS2322 `undefined` is not compatible with string/null.
      outStringNullable: "inStringUndefined",
      outStringNullableUndefined: "inStringOptional", // this is okay, optional is usually compatible with undefined
      // @ts-expect-error TS2322 `null` is not compatible with string/optional.
      outStringOptional: "inStringNullable",
      outStringOptionalNullable: "inStringUndefined",
      outStringOptionalNullableUndefined: "inString", // this is okay, it's a subset
      outStringUndefined: ObjectMapper.undefined, // Can use this convenience static method to always map to `undefined`
    });
  });

  it(`cannot use functions returning a superset of the output type`, () => {
    // Setup
    new ObjectMapper<Input, Pick<Output, "outString">>({
      // @ts-expect-error TS2322 The input value can't be `undefined` if the output value is `string`
      outString: (input) => input.inStringUndefined,
    });
  });

  it(`can convert an ObjectMapper instance into a function`, () => {
    // Setup
    const objectMapperInstance = new ObjectMapper<Input, Output>({
      outString: "inString",
      outStringNullable: ObjectMapper.null,
      outStringNullableUndefined: ObjectMapper.undefined,
      outStringOptional: "inStringOptional",
      outStringOptionalNullable: "inStringOptionalNullable",
      outStringOptionalNullableUndefined: "inStringOptionalNullableUndefined",
      outStringUndefined: "inStringUndefined",
    });

    const nextString = stringCounter();
    const input: Input = {
      inString: nextString(),
      inStringNullable: nextString(),
      inStringNullableUndefined: nextString(),
      inStringOptional: nextString(),
      inStringOptionalNullable: nextString(),
      inStringOptionalNullableUndefined: nextString(),
      inStringUndefined: nextString(),
    };
    const context = undefined;

    // Execute
    const objectMapperFunction = objectMapperInstance.toFunction();
    const output = objectMapperFunction(input, context);

    // Verify
    expect(output).toMatchSnapshot();
    expect(objectMapperFunction.schema).toBe(objectMapperInstance.schema);
  });

  describe(`nested objects`, () => {
    interface Input {
      foo: string;
    }

    interface Output {
      nested: {
        bar?: string;
      };
    }

    interface NestedInput {
      baz: string;
    }

    type NestedOutput = Output["nested"];

    it(`requires nested objects to be completely mapped`, () => {
      // Setup
      const input: Input = {
        foo: "fooValue",
      };

      const nestedObjectMapper = new ObjectMapper<NestedInput, NestedOutput>({
        bar: "baz",
      });

      const mapper = new ObjectMapper<Input, Output>({
        nested: ObjectMapper.nested(
          nestedObjectMapper,
          (input: Input) => ({ baz: input.foo }),
          ObjectMapper.undefined
        ),
      });

      // Execute
      const result = mapper.map(input, undefined);

      // Verify
      expect(result).toMatchSnapshot();
    });

    // The sub-context requires a factory function, accepting the outer input and context.
    it(`requires passing a sub-context to the nested mapper`, () => {
      // Setup
      interface NestedContext {
        capitalize: boolean;
      }

      const input: Input = {
        foo: "fooValue",
      };

      const nestedObjectMapper = new ObjectMapper<
        NestedInput,
        NestedOutput,
        NestedContext
      >({
        bar: (input, context) =>
          context.capitalize ? input.baz.toUpperCase() : input.baz,
      });

      const mapper = new ObjectMapper<Input, Output>({
        nested: ObjectMapper.nested(
          nestedObjectMapper,
          (input: Input) => ({ baz: input.foo }),
          () => ({ capitalize: true })
        ),
      });

      // Execute
      const result = mapper.map(input, undefined);

      // Verify
      expect(result).toMatchSnapshot();
    });
  });
});
