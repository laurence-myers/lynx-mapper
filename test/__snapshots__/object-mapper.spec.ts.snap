// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectMapper can convert an ObjectMapper instance into a function 1`] = `
Object {
  "outString": "dummy-string-1",
  "outStringNullable": null,
  "outStringNullableUndefined": undefined,
  "outStringOptional": "dummy-string-4",
  "outStringOptionalNullable": "dummy-string-5",
  "outStringOptionalNullableUndefined": "dummy-string-6",
  "outStringUndefined": "dummy-string-7",
}
`;

exports[`ObjectMapper can map values using functions 1`] = `
Object {
  "outString": "dummy-string-1",
  "outStringNullable": "dummy-string-2",
  "outStringNullableUndefined": "dummy-string-3",
  "outStringOptional": "dummy-string-4",
  "outStringOptionalNullable": "dummy-string-5",
  "outStringOptionalNullableUndefined": "dummy-string-6",
  "outStringUndefined": "dummy-string-7",
}
`;

exports[`ObjectMapper can map values using strings for property names on the input object 1`] = `
Object {
  "outString": "dummy-string-1",
  "outStringNullable": "dummy-string-2",
  "outStringNullableUndefined": "dummy-string-3",
  "outStringOptional": "dummy-string-4",
  "outStringOptionalNullable": "dummy-string-5",
  "outStringOptionalNullableUndefined": "dummy-string-6",
  "outStringUndefined": "dummy-string-7",
}
`;

exports[`ObjectMapper nested objects requires nested objects to be completely mapped 1`] = `
Object {
  "nested": Object {
    "bar": "fooValue",
  },
}
`;

exports[`ObjectMapper nested objects requires passing a sub-context to the nested mapper 1`] = `
Object {
  "nested": Object {
    "bar": "FOOVALUE",
  },
}
`;
